### ISSUE TITLE
NUnit runner: Handle specific non-zero exit codes

### ISSUE THREAD

**Issue by jrnail23** [cake-build/cake#517]
Right now, any time the NUnit console tool returns a non-zero exit code, the Cake NUnit runner throws a `CakeException` with the less-than-helpful error message, `NUnit: Process returned an error.`

I think this behavior is too broad, as it treats failed tests the same as an unexpected problem.
I'd like the flexibility to decide whether to blow up my build based on test failures, errors, etc.

NUnit does have meaningful exit codes ([discussed here](http://stackoverflow.com/questions/3889577/does-anyone-know-where-to-find-nunit-console-exit-codes-meanings)), so we should throw more specific exceptions as appropriate.

**Comment by jrnail23** [comment]
Also, here's the relevant source code for the NUnit runner (v2.6.4):

https://github.com/nunit/nunitv2/blob/master/src/ConsoleRunner/nunit-console/ConsoleUi.cs

According to the source ([Line 183](https://github.com/nunit/nunitv2/blob/master/src/ConsoleRunner/nunit-console/ConsoleUi.cs#L183)), it appears that a return code > 0 is the sum of the count of errors, failures, and non-runnable tests:
`returnCode = summary.Errors + summary.Failures + summary.NotRunnable;`

**Comment by patriksvensson** [comment]
I didn't know this. Excellent that you added an issue about this :+1: 

It might make sense to add a virtual member to `Tool<ToolSettings>` so that each tool can opt in to handle it's own errors.

**Comment by rprouse** [comment]
It is the same for NUnit 3. A positive return code is the number of failed tests, a negative return code is an error.

**Comment by molehillrocker** [comment]
Has anybody made any progress yet regarding this issue? We are currently facing this problem on our build server and since I have only access to the web interface, I have no clue what actually goes wrong.

I tried to hack on this a little bit this evening, but couldn't test it anymore. :astonished: 

My first idea was to redirect the overall output of NUnit to the console, since the only information we get at the moment is the above mentioned error (no other NUnit output :unamused:). This seems to be due to the ``NUnitRunner`` calling the inherited ``Tool.Run()`` method internally with two parameters only, which results in the ``ProcessSettings`` object to be ``null`` until it is instantiated as fallback in the ``Tool`` class, resulting in ``ProcessSettings.RedirectStandardOutput`` to bet set to ``false``.
Be aware that I just traced that visually, since I have my developer machine not at hands right now. :smile:

**Comment by devlead** [comment]
@molehillrocker  What CI server you using? Most output the buildlog.

You could redirect all output via the bootstrapper to a file like I do here
https://github.com/devlead/cake/blob/f/kudu/kudu.ps1

Also if you set verbosity of cake to diagnostic you will see how it's launched.

**Comment by molehillrocker** [comment]
We're using TFS 2015. Until friday I'm out of office, so I can't give you the real log output. But IIRC, it was something like:
```
Starting task Unit-Test...
*** some custom log message right before starting NUnit that was inserted for testing purposes :) ***
NUnit: Process returned an error
```
Previous tasks such as the build task were logged with lots of details, which is why I assumed that setting ``ProcessSettings.RedirectStandardOutput```may hopefully provide some more information, at least the default NUnit output (version, etc.) to see if probably already the invocation fails.

**Comment by devlead** [comment]
Actually by setting RedirectStandardOutput it won't output anything as it's redirected to an buffer that needs to be enumerated.  By default all tools should be outputted to the console.
Do you get any errors if you run it locally?
Do you have a #tool directive for nunit or do you restore via package.config?

**Comment by molehillrocker** [comment]
When it run it locally, it works fine. It is included via package.config.

**Comment by molehillrocker** [comment]
Back in the office, this is the log output from our TFS 2015:

```
<snip>
...
435 2016-01-19T14:58:25.3240665Z Finished executing task: Build-Server
436 2016-01-19T14:58:25.3240665Z ========================================
437 2016-01-19T14:58:25.3240665Z Unit-Test-Server
438 2016-01-19T14:58:25.3240665Z ========================================
439 2016-01-19T14:58:25.3240665Z Executing task: Unit-Test-Server
440 2016-01-19T14:58:25.5896943Z ##[error]An error occured when executing task 'Unit-Test-Server'.
441 2016-01-19T14:58:25.5896943Z ##[error]Error: NUnit: Process returned an error.
442 2016-01-19T14:58:25.5896943Z ----------------------------------------
443 2016-01-19T14:58:25.5896943Z Teardown
444 2016-01-19T14:58:25.5896943Z ----------------------------------------
...
</snip>
```

I'm trying to dig deeper into this now.

**Comment by devlead** [comment]
Could you turn on diagnostic logging with Cake (by passing the `-verbosity=Diagnostic` parameter or `-Verbosity Diagnostic` if you're using the powershell bootstrapper)? Then you will also get how it's launched (something like _Executing: "C:...._). Could provide more clues on what is going on.

Also which version of [Cake](https://www.nuget.org/packages/Cake) and [NUnit](https://www.nuget.org/packages/NUnit.Console) are you running?

Launching Cake with `Cake.exe -Version` will display the version (0.8.0 is the latest) and launching NUnit with `.\nunit3-console.exe --version` will display it's version (believe latest is Nuget 3.0.1 and assembly version 3.0.5. 813).

Also if you check the [NUnit3](http://cakebuild.net/dsl/nunit-v3) aliases you can pass an [NUnit3Settings](http://cakebuild.net/api/cake.common.tools.nunit/7bd0c6da), on this you can specify an [ErrorOutputFile](http://cakebuild.net/api/cake.common.tools.nunit/7bd0c6da/13910edf) and [Verbose](http://cakebuild.net/api/cake.common.tools.nunit/7bd0c6da/3f6fd78d) to get more info about what's going on.

**Comment by molehillrocker** [comment]
I installed a build of Cake from the develop branch (where the error code is printed) and found out that the process returns error code ``-2146232576`` from the NUnit process. This seems to have something to do with a .NET framework version mismatch on the build server (see [here](https://groups.google.com/forum/#!topic/nunit-discuss/wVXcc2dGK08)). We use NUnit 2.6.4 at the moment. 

In addition, on our build server, only .NET framework 4.6 seems to be installed (if I try to build with any other option than MSBuildToolVersion.NET46 resp. MSBuildToolVersion.Default, even the build fails). Also, according to the [docs](http://www.nunit.org/index.php?p=consoleCommandLine&r=2.6.4), NUnit 2.6.4 does not support higher framework versions than 4.0.

So I guess I should install either an older framework version on the build server to support NUnit 2.6.4 or migrate to NUnit 3.0.1.

**Comment by molehillrocker** [comment]
After migrating our Cake build script to use NUnit3 instead of NUnit and replacing ``<package id="NUnit.Runners" version="2.6.4" />`` by ``<package id="NUnit.Console" version="3.0.1" />`` in ``tools\packages.config``, everything works fine now. Thanks for your help, @devlead *passing a Snus over the wire* :smile:

**Comment by devlead** [comment]
@molehillrocker superb :+1:

**Comment by ChrisMaddock** [comment]
I'd be interested in taking a look at this. We wrote some new scripts for the NUnit code samples recently, and it was this limitation which meant we decided to use NUnit via the console, instead of the tool.

> It might make sense to add a virtual member to Tool<ToolSettings> so that each tool can opt in to handle it's own errors.

@patriksvensson - I'm not yet familiar with Cake's internal tool architecture, and it sounds like you have some ideas - mind posting a rough overview of what you'd expect this to look like?
